### PR TITLE
Periphery contracts gas improvements

### PR DESCRIPTION
**Pull request by garyghayrat**
Spent some time looking at different guides, but feel like I don't have enough to show for it. Particularly could use some help/suggestions as to how to improve multiple token transfers in a loop. https://github.com/ScopeLift/umbra-protocol/pull/351#discussion_r911371285

### CODE REVIEW

File: contracts-periphery/test/gas/UmbraBatchSend.gas.t.sol [hunk 1]
```diff
@@ -47,34 +53,34 @@ abstract contract UmbraBatchSendGasTest is DeployUmbraTest {
...
       router.batchSendEth{value: valueAmount}(toll, sendEth);
-
     } else if (_type == Send.TOKEN) {
       assertTrue(tokenAmount > 0);
       for (uint256 i = 0; i < numOfAddrs; i++) {
         valueAmount += toll;
         sendToken.push(UmbraBatchSend.SendToken(addrs[i], address(token), tokenAmount, pkx, ciphertext));
       }
-
+      router.batchSendTokens{value: valueAmount}(toll, sendToken);
```
Comment (garyghayrat) [inline contracts-periphery/test/gas/UmbraBatchSend.gas.t.sol:62]: Somehow missed this critical line and line 70 when initially merging this file

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 2]
```diff
@@ -57,36 +59,43 @@ contract UmbraBatchSend {
...
     }
   }
 
   function _batchSendTokens(uint256 _tollCommitment, SendToken[] calldata _params) internal {
-    for (uint256 i = 0; i < _params.length; i++) {
+    uint256 length = _params.length;
+    for (uint256 i = 0; i < length;) {
       IERC20 token = IERC20(address(_params[i].tokenAddr));
 
       SafeERC20.safeTransferFrom(token, msg.sender, address(this), _params[i].amount);
```
Comment (garyghayrat) [inline contracts-periphery/src/UmbraBatchSend.sol:74]: @mds1 Ik you've made a couple suggestion for this, but my implementations of them weren't great in previous PRs. Any particular direction you can point me to?

Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol:74]: Specifically transferring tokens each time

Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol]: Implemented SwapSummary

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 3]
```diff
@@ -24,69 +27,133 @@ contract UmbraBatchSend {
...
         _params[i].ciphertext
       );
+      unchecked {
+        i++;
+      }
+    }
+
+    uint256 summaryLength = _transferSummary.length;
+    for (uint256 j = 0; j < summaryLength; ) {
+      IERC20 token = IERC20(address(_transferSummary[j].tokenAddr));
```
Comment (apbendi) [inline contracts-periphery/src/UmbraBatchSend.sol:129]: This local variable seems to be declared but not used

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 4]
```diff
@@ -24,69 +27,133 @@ contract UmbraBatchSend {
     bytes32 ciphertext;
   }
 
+  mapping(address => uint256) internal TotalTransferAmountPerToken;
```
Comment (apbendi) [inline contracts-periphery/src/UmbraBatchSend.sol:30]: nit: should start with a lowercase 't' `totalTransferAmountPerToken`

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 5]
```diff
@@ -24,69 +27,133 @@ contract UmbraBatchSend {
...
+    TransferSummary[] calldata _transferSummary
+  ) public payable {
+    if (msg.value < _tollCommitment * _params.length) revert ValueMismatch();
+
+    for (uint256 i = 0; i < _params.length; i++) {
+      // Used later to validate total amounts are correctly provided
+      TotalTransferAmountPerToken[_params[i].tokenAddr] += _params[i].amount;
+    }
+
+    uint256 summaryLength = _transferSummary.length;
```
Comment (apbendi) [inline contracts-periphery/src/UmbraBatchSend.sol:62]: nit: this and all non-storage local variables should be prefixed with `_`

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 6]
```diff
@@ -24,69 +27,133 @@ contract UmbraBatchSend {
...
+    for (uint256 j = 0; j < summaryLength; ) {
```
Comment (apbendi) [inline contracts-periphery/src/UmbraBatchSend.sol:63]: Looks like we could re-use `i` here rather than declaring a new variable.

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 7]
```diff
@@ -24,69 +27,133 @@ contract UmbraBatchSend {
...
+      IERC20 token = IERC20(address(_transferSummary[j].tokenAddr));
+
+      if (TotalTransferAmountPerToken[_transferSummary[j].tokenAddr] != _transferSummary[j].amount)
+        revert TransferAmmountMismatch();
```
Comment (apbendi) [inline contracts-periphery/src/UmbraBatchSend.sol:67]: nit: if have a newline, I think we should include the braces

File: contracts-periphery/snapshot/.gas [hunk 8]
```diff
@@ -1,99 +1,199 @@
...
-├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
-│ Function Name   ┆ min             ┆ avg   ┆ median ┆ max   ┆ # calls │
-├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
-│ balanceOf       ┆ 534             ┆ 534   ┆ 534    ┆ 534   ┆ 4       │
-├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
-│ transfer        ┆ 29962           ┆ 29962 ┆ 29962  ┆ 29962 ┆ 8       │
-├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
-│ withdraw        ┆ 7379            ┆ 7379  ┆ 7379   ┆ 7379  ┆ 4       │
-╰─────────────────┴─────────────────┴───────┴────────┴───────┴─────────╯
+Compiling 3 files with 0.8.13
```
Comment (apbendi) [inline contracts-periphery/snapshot/.gas:1]: I'm a bit confused by this file. How does it relate to `.gas-snapshot`? Also, when I run `make snapshot-gas` on this branch, I get a big diff in this file. Is that supposed to happen? Maybe some updates to the README to explain how gas snapshotting works would be helpful.

**Review by apbendi — COMMENTED** [review]
Looks like some nice improvements @garyghayrat. Left some questions/comments/thoughts.

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 3 above)
Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol]: Removed it! https://github.com/ScopeLift/umbra-protocol/blob/b156045ea6b0d75c49dbb6500e5e9e6835131723/contracts-periphery/src/UmbraBatchSend.sol#L128-L132

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 4 above)
Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol]: gotcha https://github.com/ScopeLift/umbra-protocol/blob/b156045ea6b0d75c49dbb6500e5e9e6835131723/contracts-periphery/src/UmbraBatchSend.sol#L30

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 5 above)
Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol]: Should be good now https://github.com/ScopeLift/umbra-protocol/blob/bf77a1765f7683b3be691c8e28a230a0711c9279/contracts-periphery/src/UmbraBatchSend.sol#L62-L64

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 6 above)
Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol]: like this? It seems like we can reuse the `i` name but still have to re-declare the variable? https://github.com/ScopeLift/umbra-protocol/blob/bf77a1765f7683b3be691c8e28a230a0711c9279/contracts-periphery/src/UmbraBatchSend.sol#L57-L63

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 7 above)
Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol:67]: Yup, I confused myself for a sec there too https://github.com/ScopeLift/umbra-protocol/blob/bf77a1765f7683b3be691c8e28a230a0711c9279/contracts-periphery/src/UmbraBatchSend.sol#L66-L68

File: contracts-periphery/snapshot/.gas (hunk 8 above)
Comment (apbendi) [reply contracts-periphery/snapshot/.gas]: Context from @mds1:

```
* `forge snapshot` runs your tests and reports gas usage of each test, and creates a `.gas-snapshot` file at the root
* `forge test --gas-report` outputs a table of specific functions and their min/max/med/avg gas usage, but it just prints to the terminal so you need to manually dump it to a file
```

Comment (apbendi) [reply contracts-periphery/snapshot/.gas]: Since we're including both files, we want our `snapshot-gas` command to update both files

**Review by apbendi — APPROVED** [review]
Ok! Thanks for your incredible patience here @garyghayrat. I think we're finally in good shape 😅

File: contracts-periphery/foundry.toml [hunk 9]
```diff
@@ -1,4 +1,4 @@
-[default]
+[profile.default]
 verbosity = 3
 
 [ci]
```
Comment (mds1) [inline contracts-periphery/foundry.toml:4]:
```suggestion
[profile.ci]
```

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 10]
```diff
@@ -1,11 +1,14 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.13;
 
+import "openzeppelin-contracts/contracts/access/Ownable.sol";
 import "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
 import "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
 import "src/interface/IUmbra.sol";
+import "forge-std/console2.sol";
```
Comment (mds1) [inline contracts-periphery/src/UmbraBatchSend.sol:8]: Leftover import

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 11]
```diff
@@ -24,69 +27,131 @@ contract UmbraBatchSend {
...
-    if(msg.value != _tollCommitment * _params.length) revert ValueMismatch();
-    _batchSendTokens(_tollCommitment, _params);
+  function batchSendTokens(
+    uint256 _tollCommitment,
+    SendToken[] calldata _params,
+    TransferSummary[] calldata _transferSummary
+  ) public payable {
+    if (msg.value < _tollCommitment * _params.length) revert ValueMismatch();
+
+    for (uint256 i = 0; i < _params.length; i++) {
```
Comment (mds1) [inline contracts-periphery/src/UmbraBatchSend.sol:57]: Can use unchecked when incrementing here. If you want to dedupe a bit in all of the for loops you can do:

```solidity
function uncheckedIncrement(uint256 i) internal pure returns (uint256) {
  unchecked { return i + 1; }
}

for (uint256 i = 0; i < _params.length; i = uncheckedIncrement(i)) {
  // --- snip ---
}
```

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 12]
```diff
@@ -24,69 +27,131 @@ contract UmbraBatchSend {
...
+
+    uint256 _summaryLength = _transferSummary.length;
+    for (uint256 i = 0; i < _summaryLength; ) {
+      IERC20 _token = IERC20(address(_transferSummary[i].tokenAddr));
+
+      if (totalTransferAmountPerToken[_transferSummary[i].tokenAddr] != _transferSummary[i].amount) {
+        revert TransferAmmountMismatch();
+      }
+
+      SafeERC20.safeTransferFrom(_token, msg.sender, address(this), _transferSummary[i].amount);
```
Comment (mds1) [inline contracts-periphery/src/UmbraBatchSend.sol:70]: IMO it's a little bit cleaner/more conventional to delete the cast on L64 and change this line to:
```suggestion
      IERC20(_token).safeTransferFrom(msg.sender, address(this), _transferSummary[i].amount);
```

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 13]
```diff
@@ -24,69 +27,131 @@ contract UmbraBatchSend {
...
+    SendToken[] calldata _tokenParams,
+    TransferSummary[] calldata _transferSummary
   ) external payable {
     uint256 _totalAmount = _sumValueFromEthSends(_tollCommitment, _ethParams) + (_tollCommitment * _tokenParams.length);
-    if(msg.value != _totalAmount) revert ValueMismatch();
+    if (msg.value != _totalAmount) revert ValueMismatch();
 
     _batchSendEth(_tollCommitment, _ethParams);
-    _batchSendTokens(_tollCommitment, _tokenParams);
+    batchSendTokens(_tollCommitment, _tokenParams, _transferSummary);
```
Comment (mds1) [inline contracts-periphery/src/UmbraBatchSend.sol:90]: Why do we call the internal batch send ETH method but the external batch send tokens method?

File: contracts-periphery/foundry.toml (hunk 9 above)
Comment (garyghayrat) [reply contracts-periphery/foundry.toml:4]: added

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 10 above)
Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol:8]: removed.

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 12 above)
Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol:70]: Agreed. Changed.

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 13 above)
Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol:86]: But actually that might make less sense the more i think about it

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 14]
```diff
@@ -24,54 +26,86 @@ contract UmbraBatchSend {
...
   }
 
+  mapping(address => uint256) internal totalTransferAmountPerToken;
+  struct TransferSummary {
+    uint256 amount;
+    address tokenAddr;
+  }
+
   error ValueMismatch();
+  error TransferAmmountMismatch();
```
Comment (mds1) [inline contracts-periphery/src/UmbraBatchSend.sol:36]:
```suggestion
  error TransferAmountMismatch();
```

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 11 above)
Comment (mds1) [reply contracts-periphery/src/UmbraBatchSend.sol:56]: Looks like there's a loop in `batchSendTokens` that needs to be changed

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 13 above)
Comment (mds1) [reply contracts-periphery/src/UmbraBatchSend.sol:86]: What if we extract the eth value checks into a method that takes two arrays: one of `SendEth[]`, one of `SendToken[]`. If we're only sending ETH, we pass an empty array of `SendToken` into the helper method. Then we can reuse the same validation logic and simplify things a bit

File: contracts-periphery/.gas-snapshot [hunk 15]
```diff
@@ -0,0 +1,154 @@
+BatchSendWithTollAndWithTokenBalanceGasTest:testPostSetupState() (gas: 11857)
```
Comment (mds1) [inline contracts-periphery/.gas-snapshot:1]: _(arbitrary comment location to start a thread)_

Nit: can we delete `fs_permissions = [{ access = "read", path = "./"}]` from the `profile.ci` section? If a key is undefined, it automatically uses the default value, so we don't need to redefine it twice

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 16]
```diff
@@ -24,54 +26,86 @@ contract UmbraBatchSend {
...
+    for (uint256 i = 0; i < _params.length; i++) {
+      // Used later to validate total amounts are correctly provided
+      totalTransferAmountPerToken[_params[i].tokenAddr] += _params[i].amount;
+    }
+
+    uint256 _summaryLength = _transferSummary.length;
+    for (uint256 i = 0; i < _summaryLength; i = uncheckedIncrement(i)) {
+      IERC20 _token = IERC20(address(_transferSummary[i].tokenAddr));
+
+      if (totalTransferAmountPerToken[_transferSummary[i].tokenAddr] != _transferSummary[i].amount) {
```
Comment (mds1) [inline contracts-periphery/src/UmbraBatchSend.sol:65]: We never clear this storage mapping so won't subsequent sends revert with a mismatch here?

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 11 above)
Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol:56]: good catch, fixed

File: contracts-periphery/.gas-snapshot (hunk 15 above)
Comment (garyghayrat) [reply contracts-periphery/.gas-snapshot:1]: gotcha

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 13 above)
Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol]: Addressed here https://github.com/ScopeLift/umbra-protocol/pull/351/commits/11ee5e7c67649edb0601ab23fcc802dae11f56f8

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 17]
```diff
@@ -1,11 +1,13 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.13;
 
-import "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
-import "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+import "../lib/openzeppelin-contracts/contracts/access/Ownable.sol";
+import "../lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import "../lib/openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
```
Comment (mds1) [inline contracts-periphery/src/UmbraBatchSend.sol:6]: Why the change to relative paths?

File: contracts-periphery/src/UmbraBatchSend.sol [hunk 18]
```diff
@@ -24,54 +26,83 @@ contract UmbraBatchSend {
...
+    SendToken[] calldata _tokenParams,
+    TransferSummary[] calldata _transferSummary
   ) external payable {
-    uint256 _totalAmount = _sumValueFromEthSends(_tollCommitment, _ethParams) + (_tollCommitment * _tokenParams.length);
-    if(msg.value != _totalAmount) revert ValueMismatch();
-
-    _batchSendEth(_tollCommitment, _ethParams);
-    _batchSendTokens(_tollCommitment, _tokenParams);
+    batchSendEth(_tollCommitment, _ethParams);
+    batchSendTokens(_tollCommitment, _tokenParams, _transferSummary);
```
Comment (mds1) [inline contracts-periphery/src/UmbraBatchSend.sol:83]: This duplicates the `_checkValueFromSends` validation right? I think we can:
- Make both `function batchSend*` methods external instead of public
- Move the event emissions into the internal `function _batchSend*` methods
- In `batchSendTokens` method, it should mirror the `batchSendEth` method in that it only (1) checks the values, and (2) calls the internal version. So the for loops in `batchSendTokens` can be internal to `_batchSendTokens` instead
- All `_checkValueFromSends` calls are in the external methods, and the internal ones handle everything else

I think that should work?

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 17 above)
Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol:6]: Was getting
```
Compiler run failed
error[6275]: ParserError: Source "openzeppelin-contracts/contracts/access/Ownable.sol" not found: File not found. Searched the following locations: ".../umbra-protocol/contracts-periphery".
 --> src/UmbraBatchSend.sol:4:1:
  |
4 | import "openzeppelin-contracts/contracts/access/Ownable.sol";
```

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 18 above)
Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol]: It does. Implemented here: https://github.com/ScopeLift/umbra-protocol/pull/351/commits/fec577097bc55395bf33e0a7079845e3323a95df

File: contracts-periphery/src/UmbraBatchSend.sol (hunk 17 above)
Comment (mds1) [reply contracts-periphery/src/UmbraBatchSend.sol:6]: Huh weird, this didn't used to happen though right? Either way, IMO it's preferable to solve this with remappings since relative paths aren't as nice to work with (things break if you move around files, less clear where the file comes from)

```toml
remappings = [
  "openzeppelin-contracts/=lib/openzeppelin-contracts/contracts"
]
```

```solidity
import "openzeppelin-contracts/access/Ownable.sol";
```

Comment (mds1) [reply contracts-periphery/src/UmbraBatchSend.sol]: Reminder that you only need to define it in the default profile and the CI profile will inherit that, so we currently have:

```toml
[profile.default]
verbosity = 3
fs_permissions = [{ access = "read", path = "./"}]
remappings = ['openzeppelin-contracts/=lib/openzeppelin-contracts/contracts']
[profile.ci]
fuzz_runs = 10000
fs_permissions = [{ access = "read", path = "./"}]
remappings = ["openzeppelin-contracts/=lib/openzeppelin-contracts/contracts"]
# See more config options https://github.com/gakonst/foundry/tree/master/config
```

But we only need:
```toml
[profile.default]
verbosity = 3
fs_permissions = [{ access = "read", path = "./"}]
remappings = ['openzeppelin-contracts/=lib/openzeppelin-contracts/contracts']

[profile.ci]
fuzz_runs = 10000
```

Comment (garyghayrat) [reply contracts-periphery/src/UmbraBatchSend.sol]: gotcha, removed duplicates https://github.com/ScopeLift/umbra-protocol/pull/351/commits/88f80ca520f5b317c3fb77e0f0bcf0768b7d9a0e

**Review by mds1 — APPROVED** [review]
Thanks @garyghayrat! Can you just rebase against main, now that https://github.com/ScopeLift/umbra-protocol/pull/393 was just merged that should fix CI here

**Comment by garyghayrat** [comment]
> Thanks @garyghayrat! Can you just rebase against main, now that #393 was just merged that should fix CI here

Rebased!

**Review by apbendi — APPROVED** [review]
Looks good to me!
